### DISCUSSION TITLE
Mutating an immutable variable: temporary flaw or design decision?

### DISCUSSION THREAD

**Discussion by JoanaBLate** [vlang/v#17122] [General]
```

struct MyStruct {
mut:	
    list []string
}

fn main() {

    a := MyStruct{["aaa"]} // immutable variable
    
    mut b := a
    b.list[0] = "BBB"

    println("now first in 'a.list' is ${a.list[0]}")
    
    // output:
    // now first in 'a.list' is BBB
}
```
Sorry if it's already been discussed; I couldn't find a similar post.

**Comment by Wajinn** [comment]
The claim that such is a "flaw" is very debatable to bogus.  It was originally presented by a user who unceasingly trolls V on other websites, and is an evangelist for a competing language (who hides their affiliation), that appears not bothered by underhandedness.

The struct has been set to mutable (`mut:`).  What has been done is similar to what is present in C and other languages, by having a const pointer.  This can also be referred to as a const struct (a mutable field accessed by reference).  In Go, they avoided this by omission, where V gives users the option.  I argue this is a justifiable option that should stay in V, as it is.  The programmer should have enough sense to know what they are doing, when setting a struct to mutable or immutable.  At most, think it should only be an "alert" or warning.

If you don't want the contents of the struct to be mutable, then don't set it as so.

```
struct MyStruct {
    list []string // set to immutable
}

fn main() {

    a := MyStruct{["aaa"]} // is immutable, after assigning value
    
    mut b := a
    b.list[0] = "BBB" // mutable var b can not change values in struct of immutable var a

    println("now first in 'a.list' is ${a.list[0]}")
    
    // output:
    // error: field `list` of struct `MyStruct` is immutable
}
```

**Reply by JoanaBLate** [reply]
Look friend, I am not here for trolling or annoying. I was one of the first enthusiasts o V and first financial contributors (not on this nick; I can prove this).

Now, let's approach the technical subject. I remember long ago when I was studying Java, that I was told that all that boilerplate and verbosity, getters and setters were good for encapsulation, to protecting the fields of the object/class. And I always thought, "Bullshit! The protection only works for primitive types.". Like you said, it is how most (all?) of the other languages do.

But they use "const" which means that it is forbidden to assign a value to the variable again. Also means immutability when the variable stores a primitive value.

Now, V comes as a new language to fix the mistakes of its predecessors. Immutability is one of its iconic features, very attractive. But, unfortunately, it is flawed. And you are telling me that it is how V is supposed to work (design decision - thanks for clarifying).

I can accept that the perfect  automatic memory management that was promised never become true (it is really, really hard to achieve). I can wait years for an improved manual memory management. I accept to live forever only with the garbage collector. But I can't *TRUST* a language that shouts "Immutability is king!" and fails in the first silly example.


I really hope that you people change this design decision, because design decisions like this, send me away from V.

**Reply by Wajinn** [reply]
I apologize if it looked like I was accusing you, as you have indeed been around for a long time. I meant to reference the origins of the argument, from what I have seen in other circles.

> But they use "const" which means that it is forbidden to assign a value to the variable again. 

"Const" does not mean that, in many other languages, or at least things can be quite foggy.  Pascal is yet another language where the debate over const/constant and mutability has raged, in addition to C, so this issue has been around for a long time. 

Thus we need to be aware of **const pointer** and **const struct**.  Immutability can be relative in programming.  When a person makes the struct explicitly mutable, then they should not also expect its contents to be immutable, when accessed by reference. This is doing 2 conflicting and opposite behaviors. If we want the contents of the struct to be immutable, then we should not make them mutable (`mut`).

> Immutability is one of its iconic features, very attractive. But, unfortunately, it is flawed. 

In this case, I disagree with such an assessment.  V is giving the user a const struct option. The user can choose immutability or mutability by reference.
 
> Perfect automatic memory management that was promised...

I think we are going off topic, and it is not my perception that V promised perfection.  I look at that situation as the language providing various memory management options, that can be turned off or on. As some other languages do. GC, autofree, or manual. I expect to see improvement, over time, in these provided options.  In both features and documentation.

**Reply by JoanaBLate** [reply]
> I apologize if it looked like I was accusing you,

That's all right :-)

I see and respect your point of view.
Mine is different.
Sadly,  Iam giving up on V.

**Reply by Wajinn** [reply]
> That's all right :-)

That's good.  I didn't mean to come at you.  I do realize it looks a bit inflammatory, but was referring to the origins of the argument.

> I see and respect your point of view.
> Mine is different.

Maybe you want to state your case, as what you expect to see with your example. And why you believe the present logic is wrong.

My opinion does not represent those of the developers, but that of an interested observer.  So please don't take it as other than such.

From my point of view, in looking at this, why wouldn't a reference to a mutable struct not be changeable?  If we don't want the contents of the struct to change, we make it immutable.  The variables that are immutable, can't be changed directly, but the struct they reference can be if that struct is mutable.

1) What about this struct should mean immutability, when it explicitly is a struct that contains` mut:`? 

```
struct MyStruct {
mut:	// if we remove this then it stays immutable
    list []string
}
```


2) If we make the struct immutable, mutable a or b will fail to make changes.

```
struct MyStruct {
    list []string
}

 fn main() {

    mut a := MyStruct{["aaa"]} // does not matter if a is or is not mutable, will fail.
    mut b := a // does not matter if b is or is not mutable, will fail.
    
    b.list[0] = "BBB"
    println("now first in 'a.list' is ${a.list[0]}")
    
    // output:
    // error: field `list` of struct `MyStruct` is immutable
}
```


3. Whenever the contents of the struct are immutable, attempting indirect changes will still fail.
`b` made mutable to change contents of `a` won't work
both `a` and `b` being mutable to change the contents won't work either
```
struct MyStruct {
    list []string = ["aaa"] // is immutable
}

fn main() {

    a := MyStruct{} // does not matter if a is or is not immutable
    mut b := a
    
    b.list[0] = "BBB" // var b can not change the contents of var a, even if var a made mutable

    println("now first in 'a.list' is ${a.list[0]}")

    // output:
    // error: field `list` of struct `MyStruct` is immutable	
}
```


4) If `b` is mutable and the struct referenced is mutable, why should we expect immutability? 
What are we trying to accomplish?

`a` is referencing the struct, which is now mutable. 
Attempting `a.list[0] = "BBB"` will fail, as `a` is immutable.
But, attempting `b.list[0] = "BBB"` will succeed, as `b` is mutable.
If `b` is mutable and the struct is mutable, then `b.list[0] = "BBB"`. Why shouldn't it succeed?
If `a` and `b` are referencing a mutable struct, why shouldn't `a.list[0]` not be "BBB"?
```
struct MyStruct {
mut:	// explicitly mutable
    list []string
}

 fn main() {

    a := MyStruct{["aaa"]} // immutable variable
    mut b := a  // b is an explicitly mutable variable and so is the struct referenced
    
    b.list[0] = "BBB"

    println("now first in 'a.list' is ${a.list[0]}")
    
    // output:
    // now first in 'a.list' is BBB
}
```

**Reply by JoanaBLate** [reply]
I could present many strong arguments, but the code below is enough (I cannot express myself better than this).

I am not trying to convince anyone, just expressing my point of view.

```
struct MyStruct {
mut:
    list []string
}

fn main() {

    a := MyStruct{ ["aaa"] }
    
 // a.list[0] = "### MUTATION ###" // error: `a` is immutable, declare it with `mut` to make it mutable

    mut b:= a
    b.list[0] = "### MUTATION ###"  // OK
    
     println("now first in 'a.list' is ${a.list[0]}")
    
    // output:
    // now first in 'a.list' is ### MUTATION ### // 
    // considering the error message above, THIS IS HIGHLY *INCONSISTENT* BEHAVIOUR (BROKEN IMMUTABILITY) //
}
```

**Reply by Wajinn** [reply]
Declaring or assigning values, even as const, does not stop 3rd party objects they are pointing to from being mutated. Also, the problem in question, is only happening when variables are pointing at such 3rd party objects. Var A and Var B point to Object C.

`a` can not be given a new value directly, as immutable. But, both `a` and `b` can still reference the same 3rd party object, which could be mutated and then they both would reflect the changes.  Var B changes Object C, so Var A shows changes to Object C, even though changes could not be made to Var A directly.

To avoid such behavior, we arguably have to work with copies of the object, that have no reference to the original.  But, this can create another string of problems.  Where confusion is created from working with multiple copies of an object, increased complexity, and greater memory management issues.  

This also gets into what do we mean by making a struct mutable?  If the intention is the immutability of the values contained by the struct, then arguably they should be given outside the function.  The struct having the values and being immutable, clarifies intent (somewhat like example 3 above).  Functions, and their variables, can then only work with a copy of the immutable struct and values.

If we make the struct mutable, then point mutable variables at the 3rd party struct, then it can be assumed the intent is to allow mutability.  Changes to the 3rd party object, will be reflected by the variables pointing to it.

**Reply by impopular-guy** [reply]
I was interested in this from an observer's point of view, so was trying few things. The following makes immutable fields(non primitive) pointless. Feels like a security hazard. I do like the V lang, it is still in dev so i think anything can happen.
```v
struct MyStruct {
	list []string
}

fn main() {
	a := MyStruct{['AAA']}
	println('before: ${a}')
	
	mut l := a.list
	l[0] = 'DDD'
	println('after: ${a}')
}
```
```
Output:
before: MyStruct{
    list: ['AAA']
}
after: MyStruct{
    list: ['DDD']
}
```

**Reply by Wajinn** [reply]
> ```
> mut l := a.list
> ```

What version of V are you using?  
That line shows-  `notice: use mut array2 := array1.clone() instead of mut array2 := array1 (or use unsafe)`

The fix,` mut l := a.list.clone()`, would then work.  As you are dealing with a copy.

**Reply by impopular-guy** [reply]
@Wajinn I did it in play.vlang.io
If it puts a notice then it is great thing.

I think i clearly understand the immutability. From a hardware point of view, the data in memory pointed by that variable will remain immutable, doesn't matter what kind of data it is.

edit: it does put notice on local machine with latest v

**Reply by Wajinn** [reply]
> If it puts a notice then it is great thing.

I think that is what we need, a notice in such ambiguous or debatable usage cases.

**Reply by larpon** [reply]
I think this might just be a bug. In other cases `mut b := a` (where `a` is immutable) it requires an `unsafe {}` block to do the declaration of `b` .
It works like that for consts AFAIK

**Reply by Wajinn** [reply]
> I think this might just be a bug. In other cases `mut b := a` (where `a` is immutable) it requires an `unsafe {}` block to do the declaration of `b` . 

I kind of agree with that, the bug label, and resolving it with unsafe {} (or notice). Also, discerning intent in such cases is difficult. If the intent is something like const pointer or const struct, putting it into an unsafe {} block or displaying a notice should be fine.

However, mostly side with just a notice (as shown earlier), because a user having mutable variables referencing a mutable struct, should not expect "all around" immutability. Looks like opposing contradictions with user chosen actions versus expectations. Maybe another solution for resolving such confusion or contradictions between actions and expectations is putting something about it or clarifying it in the documentation.

There is something about such a conflicted expectation, that bothers me a bit. In those situations, I see it as helpful that the compiler gives a notice so that we can think about what we might be trying to do.  It's hard for me to decide if such actions should really be designated as unsafe or should be seen as consequences of user chosen options (inappropriate use of mut or ambiguous intent).

**Reply by JoanaBLate** [reply]
Excellent! 

Now that you accept that mutating an immutable variable is not simply "how V is supposed to work" we can move to the next phase.

Next phase:

If the user has the intention of doing something stupid, then... 

Wait! How can you be sure that the user ***HAS*** the intention? Maybe he is doing a painful mistake/typo that will create a VERY HARD TO FIND BUG. So he must be forbidden!!!

In case, he really has the intention. You forbid him and tell him to write CORRECT CODE by making the variable 'a'  mutable!

There is no need to increase complexity in V creating exceptions for corner cases.

EDIT: if the compiler warns, it solves the part about "has the intention"; maybe OK at compile time. But is bad for who reads the code later. Mutating an immutable variable should not be admited.

**Reply by Wajinn** [reply]
> Now that you accept that mutating an immutable variable is not simply "how V is supposed to work" we can move to the next phase.

That is not quite what I believe is going on or how I see it.  

While the variable itself might be designated as immutable, that doesn't necessarily apply to the object it is pointing to/referencing or other variables that may also point to/reference that same object.  So there are different situations or conditions that the variable may be in.

I think there is a distinction or misconception about immutability, which might be at the heart of misunderstandings or the way that the narrative is being framed.  Though, it should be noted, that competitors of V appeared to have purposely mangled and twisted the narrative versus having an actual misunderstanding.  Hopefully, there can be more clarity on what is actually going on.

**1)** The variable may hold no references to objects outside of the function.

In that case, "simplistic" immutability applies.  In this case, V never had a problem.

**2)** The variable may point to/reference an object outside of the function.

In this case, the immutable concept becomes cloudy **for all programming languages**.  This is because we may want to mutate what is considered a constant or default values.  Thus concepts such as const pointer and const struct.  

Some languages, such as Go, completely ran from const struct and instead use "helper" functions to give initial default values of the struct.  It's a kind of workable cheap trick of Go (which can be used in V too).  In V, however, we are empowered with the choice between immutability or mutability by reference and the choice of having structs with default values. 

As many other languages don't handle immutability explicitly or obviously, people's experiences in those languages may not apply or distort their perception of what is going on in V.  They may think in terms of simplistic immutability, like "on" or "off", where there is more going on.  

When we choose to make an object outside of the function mutable, we may need to temper our expectations over immutability and proceed with caution, in order to get what we actually want.  

**3)** If we are working with a mutable object outside of the function, it must be decided if we want to interact with it directly or with a copy.

This is where the intent of the user comes in.  Does the user want to work with a copy or directly modify objects outside of the function by reference?

This is where conflicted intent and expectations can come in.  The user has mutable objects, is mixing immutable variables with mutable variables (by possible inappropriate use of `mut`), but expects "full" immutability.

> You forbid him and tell him to write CORRECT CODE by making the variable 'a' mutable!

This is not what I've stated nor are referring to.  In no way is the user being "forbidden" from doing anything.  If the user's expectation is "full" immutability, then they should be cautious or understand what they are doing when mixing in mutable objects and mutable variables (by using` mut`).

The compiler is giving a notice, when the user is trying to assign or mixing the immutable with mutable.  If the user is not working with a copy of the object inside the function, then it is being considered to use `unsafe {}` for when they are attempting to modify an outside object directly (by reference).  I'm actually kind of against or undecided about such an `unsafe {}` designation.  I believe the user should understand what they are doing, not be so careless, or ambiguous.  If they are, then suffer the consequences of not getting what they want.  That's how we learn.

> There is no need to increase complexity in V creating exceptions for corner cases.

Again, I'm not sure where we are going with this.  Team V is handling an issue as presented to them.  V also provided the ability to choose immutable or mutable from the beginning.  This is a core feature of the language that is being addressed, to increase usability for users.  And all programming languages, in one way or another, must deal with or address immutability versus mutability.  There is no true increase in complexity, when such a feature and concept already exists.  It's just an advanced topic that all programmers and languages must confront at some point and in some way.

**Reply by Ddiidev** [reply]
I don't have as deep a knowledge as my colleagues, but if I may point something out.
1. Simpler to just print a warn as a potential error.
2. apply a new keyword, eg ref
`mut b := ref a // mut a`
// That way I would be making it clear that I want a reference and not a copy of the struct.

Makes sense?
_If you couldn't explain why doesn't it make sense?_

**Reply by JoanaBLate** [reply]
Where I wrote 

"You forbid him and tell him to write CORRECT CODE by making the variable 'a' mutable!" I

please read

"V should forbid him..."; (my fault when writing the message).

About everything else, I was not expecting a long and deep discussion. 

I just wanted to point something that I found very important.

I just read the message from JalonSolov.

I will not reply anymore.
 
Thanks for your attention ;)

**Reply by Wajinn** [reply]
> 1. Simpler to just print a warn as a potential error.

I lean towards such, as in a notice to the user.

> 2. apply a new keyword, eg ref

Don't know about using the `ref keyword`, but maybe some form of the concept could work.  Maybe some new function that forces mutability of the object to be the same as the variable it's assigned to, inside of a function.  

So if part of the problem is awareness of or dealing with 2 levels of mutability, the object's mutability versus the variable's mutability it's assigned to, then maybe something like, `a := MyStruct{["aaa"]}.sync()`.  So now mutability is synchronized on assignment and it's ensured that both are the same, and that it's a copy.

Yet another way that might work, declare and assign the struct inside the function.  Instead of confusion about the mutability of an object outside the function versus the mutability of the variable its assigned to inside the function, the user can declare and assign the struct inside the function. 

```
fn foo(){
    mut my_struct := struct {
        list []string
        }
}
```

**Comment by JalonSolov** [comment]
Perhaps it would be better to just ask the right person, to get the definitive answer?

@medvednikov can you answer the original question posed in this thread?

**Comment by medvednikov** [comment]
This is a bug, just like the `mut x := my_const` which was fixed recently.

Here things are complicated by the array field, so it'll be trickier to fix.

**Comment by JalonSolov** [comment]
@JoanaBLate please open an issue on this so it can be fixed.

**Reply by JoanaBLate** [reply]
OK

**Reply by JalonSolov** [reply]
Thank you!
